File: swift/Relativedelta.swift
import Foundation
 
class Relativedelta {
  var year: Int? = nil
  var month: Int? = nil
  var day: Int? = nil
  var hour: Int? = nil
  var minute: Int? = nil
  var second: Int? = nil
 
  var years = 0
  var months = 0
  var days = 0
  var hours = 0
  var minutes = 0
  var seconds = 0
  var leapdays = 0
 
  init(baseDay: Date, day: Date) throws {
 
    // if (day.compare(baseDay) == .orderedAscending) {
    if day < baseDay {
      throw NSError(domain: "Illegal Argument Exception", code: -1, userInfo: nil)
    }
 
    let cal = Calendar(identifier: .gregorian)
    var months =
      (cal.component(.year, from: day) - cal.component(.year, from: baseDay)) * 12
      + (cal.component(.month, from: day) - cal.component(.month, from: baseDay))
    self.setMonths(months: months)
 
    var dtm = self.getDTM(other: baseDay)
 
    let gt: (Date, Date) -> ComparisonResult = { $0.compare($1) }
    let lt: (Date, Date) -> ComparisonResult = { $1.compare($0) }
    var compare: (Date, Date) -> ComparisonResult
    var increment = 0
    // if (day.compare(baseDay) == .orderedAscending) {
    if day < baseDay {
      compare = gt
      increment = 1
    } else {
      compare = lt
      increment = -1
    }
 
    while compare(day, dtm) == .orderedDescending {
      months += increment
      self.setMonths(months: months)
      dtm = self.getDTM(other: baseDay)
    }
 
    // let interval = DateInterval(start: day, end: dtm)
    let interval = DateInterval(start: dtm, end: day)
    self.seconds = Int(interval.duration)
 
    if self.seconds > 59 {
      let s = self.sign(Double(self.seconds))
      let div = self.seconds * s / 60
      let mod = (self.seconds * s) % 60
      self.seconds = mod * s
      self.minutes += div * s
    }
    if self.minutes > 59 {
      let s = self.sign(Double(self.minutes))
      let div = self.minutes * s / 60
      let mod = (self.minutes * s) % 60
      self.minutes = mod * s
      self.hours += div * s
    }
    if self.hours > 23 {
      let s = self.sign(Double(self.hours))
      let div = self.hours * s / 24
      let mod = (self.hours * s) % 24
      self.hours = mod * s
      self.days += div * s
    }
    if self.months > 11 {
      let s = self.sign(Double(self.months))
      let div = self.months * s / 12
      let mod = (self.months * s) % 12
      self.months = mod * s
      self.years += div * s
    }
 
    self.year = self.years
    self.month = self.months
    self.day = self.days
    self.hour = self.hours
    self.minute = self.minutes
    self.second = self.seconds
  }
 
  private func setMonths(months: Int) {
    self.months = months
    if self.months > 11 {
      let s = self.sign(Double(self.months))
      let div = self.months * s / 12
      let mod = (self.months * s) % 12
      self.months = mod * s
      self.years = div * s
    } else {
      self.years = 0
    }
  }
 
  private func getDTM(other: Date) -> Date {
    let cal = Calendar(identifier: .gregorian)
    var y = cal.component(.year, from: other) + self.years
    if let _y = self.year {
      y = _y + self.years
    }
    var m = self.month != nil ? self.month! : cal.component(.month, from: other)
    if self.months > 0 {
      m = m + self.months
      if m > 12 {
        y += 1
        m -= 12
      } else if m < 1 {
        y -= 1
        m += 12
      }
    }
    let _d = self.day != nil ? self.day! : cal.component(.day, from: other)
    var comps = DateComponents()
    comps.year = y
    comps.month = m
    var dayInMonth = 0
    if let date = cal.date(from: comps) {
      if let rng = cal.range(of: .day, in: .month, for: date) {
        dayInMonth = rng.upperBound - 1
      }
    }
 
    let d = min(dayInMonth, _d)
 
    let repl = (
      year: y,
      month: m,
      day: d,
      hour: self.hour != nil ? self.hour! : cal.component(.hour, from: other),
      minute: self.minute != nil ? self.minute! : cal.component(.minute, from: other),
      second: self.second != nil ? self.second! : cal.component(.second, from: other)
    )
    var days = self.days
    if self.leapdays > 0 && m > 2 && self.isYearLeapYear(y) {
      days += self.leapdays
    }
    //let ret = cal.setY(repl.year, M:repl.month, d:repl.day, h:repl.hour, m:repl.minute, s:repl.second)!
    let ret = cal.setY(repl.year, repl.month, repl.day, repl.hour, repl.minute, repl.second)!
    comps = DateComponents()
    comps.day = days
    comps.hour = self.hours
    comps.minute = self.minutes
    comps.second = self.seconds
    return cal.date(byAdding: comps, to: ret)!
  }
 
  private func sign(_ x: Double) -> Int {
    return (x < 0) ? -1 : (x > 0) ? +1 : 0
  }
 
  private func isYearLeapYear(_ year: Int) -> Bool {
    var result = false
 
    if year % 4 == 0 {
      if year % 100 == 0 {
        result = year % 400 == 0
      } else {
        return true
      }
    }
 
    return result
  }
}

File: c/relativedelta.c
#include "relativedelta.h"
 
DATES Relativedelta(struct tm baseDay, struct tm day) {
 
  struct tm dtm = {0};
  Compare compare = NULL;
  int increment = 0;
  int duration = 0;
  int s = 0;
  int div = 0;
  int mod = 0;
 
  DATES dates = {
      .year = -1,
      .month = -1,
      .day = -1,
      .hour = -1,
      .minute = -1,
      .second = -1,
      .years = 0,
      .months = 0,
      .days = 0,
      .hours = 0,
      .minutes = 0,
      .seconds = 0,
  };
 
  if (lt(&day, &baseDay) > 0) {
    errno = EINVAL;
    perror("Error");
    exit(1);
  }
 
  int months = ((day.tm_year + 1900) - (baseDay.tm_year + 1900)) * 12 +
               ((day.tm_mon + 1) - (baseDay.tm_mon + 1));
 
  setMonths(months, &dates);
 
  dtm = getDTM(baseDay, &dates);
 
  if (lt(&day, &baseDay) > 0) {
    compare = gt;
    increment = 1;
  } else {
    compare = lt;
    increment = -1;
  }
 
  while (compare(&day, &dtm) > 0) {
    months += increment;
    setMonths(months, &dates);
    dtm = getDTM(baseDay, &dates);
  }
 
  duration = difftime(mktime(&day), mktime(&dtm));
  dates.seconds = duration;
 
  if (dates.seconds > 59) {
    s = signbit(dates.seconds) == 0 ? 1 : 0;
    div = (dates.seconds * s) / 60;
    mod = (dates.seconds * s) % 60;
    dates.seconds = mod * s;
    dates.minutes += div * s;
  }
  if (dates.minutes > 59) {
    s = signbit(dates.minutes) == 0 ? 1 : 0;
    div = (dates.minutes * s) / 60;
    mod = (dates.minutes * s) % 60;
    dates.minutes = mod * s;
    dates.hours += div * s;
  }
  if (dates.hours > 23) {
    s = signbit(dates.hours) == 0 ? 1 : 0;
    div = (dates.hours * s) / 24;
    mod = (dates.hours * s) % 24;
    dates.hours = mod * s;
    dates.days += div * s;
  }
  if (dates.hours > 11) {
    s = signbit(dates.months) == 0 ? 1 : 0;
    div = (dates.months * s) / 12;
    mod = (dates.months * s) % 12;
    dates.months = mod * s;
    dates.years += div * s;
  }
 
  dates.year = dates.years;
  dates.month = dates.months;
  dates.day = dates.days;
  dates.hour = dates.hours;
  dates.minute = dates.minutes;
  dates.second = dates.seconds;
 
  return dates;
}
 
struct tm getDate(int y, int m, int d, int hour, int min, int sec) {
  struct tm ts = {0};
  time_t new_time;
  struct tm *new_ts = {0};
 
  ts.tm_year = y - 1900;
  ts.tm_mon = m - 1;
  ts.tm_mday = d;
  ts.tm_hour = hour;
  ts.tm_min = min;
  ts.tm_sec = sec;
 
  new_time = mktime(&ts);
  new_ts = localtime(&new_time);
  return *new_ts;
}
 
void setMonths(int months, LPDATES lpDates) {
  lpDates->months = months;
  if (lpDates->months > 11) {
    int s = signbit(lpDates->months) == 0 ? 1 : 0;
    int div = (lpDates->months * s) / 12;
    int mod = (lpDates->months * s) % 12;
    lpDates->months = mod * s;
    lpDates->years = div * s;
  } else {
    lpDates->years = 0;
  }
}
 
struct tm getDTM(struct tm other, LPDATES lpDates) {
  struct tm repl = {0};
  struct tm otherTm = {0};
  struct tm newTm = {0};
  int y = 0;
  if (lpDates->year != -1) {
    y = lpDates->year + lpDates->years;
  } else if (other.tm_year > 0) {
    y = other.tm_year + lpDates->years;
  }
  int m = lpDates->month != -1 ? lpDates->month : other.tm_mon + 1;
  if (lpDates->months > 0) {
    m += lpDates->months;
    if (m > 12) {
      y += 1;
      m -= 12;
    } else if (m < 1) {
      y -= 1;
      m += 12;
    }
  }
  int _d = lpDates->day != -1 ? lpDates->day : other.tm_mday;
  int d = MIN(lengthOfMonth(y, m), _d);
 
  repl.tm_year = y;
  repl.tm_mon = m;
  repl.tm_mday = d;
  repl.tm_hour = lpDates->hour;
  repl.tm_min = lpDates->minute;
  repl.tm_sec = lpDates->second;
 
  int days = lpDates->days;
  if (lpDates->leapdays > 0 && m > 2 && IsLeapYear(y) > 0) {
    days += lpDates->leapdays;
  }
 
  otherTm.tm_year = other.tm_year;
  otherTm.tm_mon = other.tm_mon;
  otherTm.tm_mday = other.tm_mday;
  otherTm.tm_hour = other.tm_hour;
  otherTm.tm_min = other.tm_min;
  otherTm.tm_sec = other.tm_sec;
 
  newTm.tm_year = repl.tm_year != -1 ? repl.tm_year : otherTm.tm_year;
  newTm.tm_mon = repl.tm_mon != -1 ? repl.tm_mon : otherTm.tm_mon;
  newTm.tm_mday = repl.tm_mday != -1 ? repl.tm_mday : otherTm.tm_mday;
  newTm.tm_hour = repl.tm_hour != -1 ? repl.tm_hour : otherTm.tm_hour;
  newTm.tm_min = repl.tm_min != -1 ? repl.tm_min : otherTm.tm_min;
  newTm.tm_sec = repl.tm_sec != -1 ? repl.tm_sec : otherTm.tm_sec;
 
  return getDate(newTm.tm_year + 1900, newTm.tm_mon, newTm.tm_mday + days,
                 newTm.tm_hour + lpDates->hours,
                 newTm.tm_min + lpDates->minutes,
                 newTm.tm_sec + lpDates->seconds);
}
 
int IsLeapYear(int year) {
  if (((year % 4 == 0) && (year % 100 != 0)) || year % 400 == 0) {
    return 1;
  } else {
    return 0;
  }
}
 
int lengthOfMonth(int y, int m) {
  int leap;
  int lmdays[] = {31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
  int mdays[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
 
  leap = IsLeapYear(y);
  if (leap == 1) {
    return lmdays[m - 1];
  } else {
    return mdays[m - 1];
  }
}
 
int gt(struct tm *_a, struct tm *_b) {
  time_t a = mktime(_a);
  time_t b = mktime(_b);
  if (a > b)
    return 1;
  else
    return -1;
}
 
int lt(struct tm *_a, struct tm *_b) {
  time_t a = mktime(_a);
  time_t b = mktime(_b);
  if (a < b)
    return 1;
  else
    return -1;
}
 
void printDate(struct tm *ts) {
  char buf[255] = {0};
  strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", ts);
  printf("%s\n", buf);
}
 
char *getPrintDate(struct tm *ts) {
  static char buf[255] = {0};
  memset(buf, 0, sizeof(buf));
  strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", ts);
  return buf;
}
 
void printElapsedDateTime(struct tm baseDay, struct tm day) {
  DATES dates = Relativedelta(baseDay, day);
  printf("%s: %d 年と %d ヶ月と %d 日と %d 時間と %d 分と %d 秒\n",
         getPrintDate(&day), dates.year, dates.month, dates.day, dates.hour,
         dates.minute, dates.second);
}
